File: server/src/models/UserProduct.ts
import {
  Table,
  Model,
  Column,
  IsUUID,
  PrimaryKey,
  Default,
  DataType,
  BelongsTo,
  ForeignKey,
  AllowNull,
  Is,
} from "sequelize-typescript";
import { Field, ObjectType } from "type-graphql";
import { Product } from "./Product";
import { User } from "./User";

@ObjectType()
@Table({
  timestamps: true,
  underscored: true,
})
export class UserProduct extends Model {
  @Field()
  @IsUUID(4)
  @PrimaryKey
  @Default(DataType.UUIDV4)
  @Column
  id: string;

  @IsUUID(4)
  @ForeignKey(() => User)
  @AllowNull(false)
  @Column
  userId: string;

  @BelongsTo(() => User)
  user: User;

  @IsUUID(4)
  @ForeignKey(() => Product)
  @AllowNull(false)
  @Column
  productId: string;

  @BelongsTo(() => Product)
  product: Product;

  @Field()
  @Default(1)
  @Is("Positive", (value) => {
    if (value <= 0) {
      throw new Error(`"${value}" must be greater than 0.`);
    }
  })
  @AllowNull(false)
  @Column
  count: number;

  @Field()
  createdAt: Date;

  @Field()
  updatedAt: Date;
}

File: go-server/internal/models/user_product.go
package models

import (
	"fmt"

	"gorm.io/gorm"
)

func init() {
	registerModel(&UserProduct{})
}

// UserProduct is a cart line item: one row per (user, product) pair. Only
// count/createdAt/updatedAt carry @Field() on the Node model — userId and
// productId are DB-only, and "product" is a FieldResolver
// (CartResolver.Product in graph/cart.resolvers.go), so this struct field
// exists for the foreign key/cascade constraint, not for GraphQL autobind
// (product is forced to a resolver in gqlgen.yml).
type UserProduct struct {
	Base

	UserID string `gorm:"not null;index" json:"-"`
	User   *User  `gorm:"foreignKey:UserID;constraint:OnDelete:CASCADE" json:"-"`

	ProductID string   `gorm:"not null;index" json:"-"`
	Product   *Product `gorm:"foreignKey:ProductID;constraint:OnDelete:CASCADE" json:"-"`

	Count int `gorm:"not null;default:1" json:"count"`
}

// BeforeSave mirrors the `@Default(1)` + `@Is("Positive", ...)` pair on
// count. Sequelize applies a declared default to the instance before
// running validators, so an unset count becomes 1 first, then passes
// validation — applying only the GORM column default (which only takes
// effect in the SQL GORM emits) would leave the in-memory zero value to
// fail this check first.
func (up *UserProduct) BeforeSave(tx *gorm.DB) error {
	if up.Count == 0 {
		up.Count = 1
	}
	if up.Count <= 0 {
		return fmt.Errorf(`"%d" must be greater than 0.`, up.Count)
	}
	return nil
}
